server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,433 @@
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final MasterHistoryService masterHistoryService;
+    /**
+     * This is the amount of time we use to make the initial decision -- have we seen a master node in the very recent past?
+     */
+    private final TimeValue veryRecentPast;
+    /**
+     * This is the number of times that it is OK for the master history to show a transition from a non-null master to a null master before
+     * it starts impacting the health status.
+     */
+    private final int acceptableNullTransitions;
+    /**
+     * This is the number of times that it is OK for the master history to show a transition one non-null master to a different non-null
+     * master before it starts impacting the health status.
+     */
+    private final int acceptableIdentityChanges;
+
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    // This is the default amount of time we look back to see if we have had a master at all, before moving on with other checks
+    private static final TimeValue DEFAULT_VERY_RECENT_PAST = new TimeValue(30, TimeUnit.SECONDS);
+    private static final TimeValue SMALLEST_ALLOWED_VERY_RECENT_PAST = new TimeValue(1, TimeUnit.SECONDS);
+
+    // This is the default number of times that it is OK to have a master go null. Any more than this will be reported as a problem
+    private static final int DEFAULT_ACCEPTABLE_NULL_TRANSITIONS = 3;
+    private static final int SMALLEST_ALLOWED_ACCEPTABLE_NULL_TRANSITIONS = 0;
+
+    // This is the default number of times that it is OK to have a master change identity. Any more than this will be reported as a problem
+    private static final int DEFAULT_ACCEPTABLE_IDENTITY_CHANGES = 3;
+    private static final int SMALLEST_ALLOWED_ACCEPTABLE_IDENTITY_CHANGES = 0;
+
+    // Keys for the details map:
+    private static final String DETAILS_CURRENT_MASTER = "current_master";
+    private static final String DETAILS_RECENT_MASTERS = "recent_masters";
+    private static final String DETAILS_EXCEPTION_FETCHING_HISTORY = "exception_fetching_history";
+    private static final String DETAILS_EXCEPTION_FETCHING_HISTORY_STACK_TRACE = "exception_fetching_history_stack_trace";
+
+    // Impacts of having an unstable master:
+    private static final String UNSTABLE_MASTER_INGEST_IMPACT = "The cluster cannot create, delete, or rebalance indices, and cannot "
+        + "insert or update documents.";
+    private static final String UNSTABLE_MASTER_DEPLOYMENT_MANAGEMENT_IMPACT = "Scheduled tasks such as Watcher, ILM, and SLM will not "
+        + "work. The _cat APIs will not work.";
+    private static final String UNSTABLE_MASTER_BACKUP_IMPACT = "Snapshot and restore will not work.";
+
+    public static final Setting<TimeValue> VERY_RECENT_PAST_SETTING = Setting.timeSetting(
```

Comment:
Is this history not already configured for the master service? I wouldn't think we'd need this configuration again? (as the master history service has a definition of "recent" already via `maxHistoryAge` https://github.com/elastic/elasticsearch/blob/f69c7396d9334d16c2b6a1614d5f32fb8a250d1c/server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java#L37 )

Do we envision these 2 settings to ever diverge? What would the consequence of that be and why would we want to?

Reply:
Ah, that's correct, my bad. Thanks for the clarification!